ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/ApplyMigrationCommand.java

```diff
@@ -56,7 +77,188 @@
 
   @Override
   protected int command() {
-    throw new UnsupportedOperationException();
+    if (!validateConfigFilePresent()) {
+      return 1;
+    }
+
+    final MigrationConfig config;
+    try {
+      config = MigrationConfig.load(configFile);
+    } catch (KsqlException | MigrationException e) {
+      LOGGER.error(e.getMessage());
+      return 1;
+    }
+
+    return command(
+        config,
+        MigrationsUtil::getKsqlClient,
+        getMigrationsDirFromConfigFile(configFile),
+        Clock.systemDefaultZone()
+    );
+  }
+
+  @VisibleForTesting
+  int command(
+      final MigrationConfig config,
+      final Function<MigrationConfig, Client> clientSupplier,
+      final String migrationsDir,
+      final Clock clock
+  ) {
+    final Client ksqlClient;
+    try {
+      ksqlClient = clientSupplier.apply(config);
+    } catch (MigrationException e) {
+      LOGGER.error(e.getMessage());
+      return 1;
+    }
+
+    if (ValidateMigrationsCommand.validate(config, migrationsDir, ksqlClient)
+        && apply(config, ksqlClient, migrationsDir, clock)) {
+      ksqlClient.close();
+      return 0;
+    } else {
+      ksqlClient.close();
+      return 1;
+    }
+  }
+
+  private boolean apply(
+      final MigrationConfig config,
+      final Client ksqlClient,
+      final String migrationsDir,
+      final Clock clock
+  ) {
+    final int start = getStartVersion(config, ksqlClient);
+    final List<Migration> migrations;
+    LOGGER.info("Loading migration files");
+
+    if (next) {
+      migrations = getAllMigrations(start, start, migrationsDir);
```

Comment:
As above, this logic doesn't quite work since the next version does not have to be consecutive. The "next" version can use the same logic as "all" to find future migration files (from the latest applied version) but should only apply the first version found. 

We might also need to check whether the version(s) to be applied have already been applied or not. It's possible that we went "backwards" in version due to an out of order migration. We should check with @colinhicks what the expected user experience is here. Specifically, suppose a user applies migrations with versions 1, 2, and 6, then applies version 3 out of order. If the user then tries to apply the "next" migration, do we apply the migration with version 4 or 7?

Reply:
The design doc suggests this option:

```
  --out-of-order     Allow out of order migrations. Only applicable for `apply`. This does not affect the behavior of `next` but will affect `all` and `until`.``
```

Reply:
I think it's still ambiguous. For example, if you apply 1 then 2 then 6 then 3, does "CURRENT" now point to 6 or 3? I think it should point to 3, and that the "previous" state linked to from 3 should be 6. If that's the case then "next" from the "CURRENT" of 3 is still 4.

Reply:
I agree with @vcrfxia above.

> Specifically, suppose a user applies migrations with versions 1, 2, and 6, then applies version 3 out of order. If the user then tries to apply the "next" migration, do we apply the migration with version 4 or 7?

As a user, I would expect version 4.

> I think it's still ambiguous. For example, if you apply 1 then 2 then 6 then 3, does "CURRENT" now point to 6 or 3? I think it should point to 3, and that the "previous" state linked to from 3 should be 6. If that's the case then "next" from the "CURRENT" of 3 is still 4.

This also sounds good to me.

Reply:
@colinhicks Won't that behavior cause confusions with users? If I migrated to 6, and for some reason I applied an out-of-order migration 3, why would we continue from 4 after that? Btw, I don't understand out-of-order migrations yet, but if users use it, then those should be in special cases, i.e. to apply a migration that they forgot (oops!). I would expect the tool continues applying migrations from the highest version migrated, in this case next one should be 7. 

Btw, do we need to support out-of-order migrations?

Reply:
@spena can you clarify the line here?

>   --out-of-order     Allow out of order migrations. Only applicable for `apply`. This does not affect the behavior of `next` but will affect `all` and `until`.``

I think I might be misinterpreting it. Are we actually saying that `--out-of-order` handles the case where the current migration is 6 and a new migration, 4, was added late, after we had applied 1, 2, 3, 5, and 6?

Reply:
> Are we actually saying that --out-of-order handles the case where the current migration is 6 and a new migration, 4, was added late, after we had applied 1, 2, 3, 5, and 6?

I added that line to the KLIP. My intent in saying "This does not affect the behavior of next" was that if the current migration is 6, then `next` will not apply 4 even if `--out-of-order` is set. Upon thinking about it more, I think we should either disallow setting `--out-of-order` for the `next` mode (to avoid confusion) or allow it with the semantics of finding the first, unapplied migration, which may be before the "current" version.

Reply:
> Btw, do we need to support out-of-order migrations?

We've already decided to not support this for the first version of the migrations tool, so I assume you're asking about the long-term plan instead? I thought it was a useful feature to have in the long run but maybe not if it's generating so much confusion. I'll let @colinhicks make the call here.

A related question is whether we want to allow non-consecutive migration versions (e.g., as a user can I apply version 1 and then version 3 without applying version 2). Doing so without supporting out of order migrations means I'll never be able to apply version 2 (on this cluster). This may still be valuable to users who, for example, start using the migrations tool from a non-clean slate and create migration 1 to capture the current state but cannot apply it (because those streams and tables already exist), and want to start from migration 2. Does this make sense, or would supporting non-consecutive migrations introduce more confusion than it's worth, if we don't also support out of order migrations (at least in v1 of the migrations tool)?

Reply:
Thanks for the explanation.

> I think we should either disallow setting --out-of-order for the next mode (to avoid confusion) or allow it with the semantics of finding the first, unapplied migration

Based on the conversation above, my vote is to disallow setting it for `next` in order to avoid confusion :). How do others feel about this?

Reply:
I do think we want to keep the option open to support other out-of-order migration scenarios in future versions. Ideally we can be informed by user feedback when considering cases.